ros_packages/voice_assistant/voice_assistant/assistant.py

```diff
@@ -1,285 +1,385 @@
+        future: Future = self.record_audio_client.send_goal_async(goal, feedback_callback)
+        return self.digest_goal_handle_future(future, result_callback)
+
+    def chat(self, 
+             text: str, 
+             chat_id: str, 
+             on_sentence_received: Callable[[str], None] = None, 
+             on_final_sentence_received: Callable[[str], None] = None) -> Callable[[], None]:
+        
+        goal = Chat.Goal()
+        goal.text = text
+        goal.chat_id = chat_id
+        feedback_callback = None if on_sentence_received is None else lambda msg: on_sentence_received(msg.feedback.sentence)
+        result_callback = None if on_final_sentence_received is None else lambda result: on_final_sentence_received(result.rest)
+        future: Future = self.chat_client.send_goal_async(goal, feedback_callback)
+        return self.digest_goal_handle_future(future, result_callback)
+
+    def play_audio_from_file(self, 
+                             filepath: str, 
+                             on_stopped_playing: Callable[[], None] = None):
+        
+        request = PlayAudioFromFile.Request()
+        request.filepath = filepath
+        request.join = on_stopped_playing is not None
+        future: Future = self.play_audio_from_file_client.call_async(request)
+        if request.join: future.add_done_callback(lambda _ : on_stopped_playing())
+
+    def play_audio_from_speech(self, 
+                               speech: str, 
+                               gender: str, 
+                               language: str, 
+                               on_stopped_playing: Callable[[], None] = None):
         
+        request = PlayAudioFromSpeech.Request()
+        request.speech = speech
+        request.gender = gender
+        request.language = language
+        request.join = on_stopped_playing is not None
+        future: Future = self.play_audio_from_speech_client.call_async(request)
+        if request.join: future.add_done_callback(lambda _ : on_stopped_playing())
 
-    def set_voice_assistant_state(self, request: SetVoiceAssistantState.Request, response: SetVoiceAssistantState.Response):
 
+    # serivce callbacks -----------------------------------------------------------------
+
+
+    def get_voice_assistant_state(self, _: GetVoiceAssistantState.Request, response: GetVoiceAssistantState.Response) -> GetVoiceAssistantState.Response:
+
+        response.voice_assistant_state = self.state
+        return response
+
+
+
+    def set_voice_assistant_state(self, request: SetVoiceAssistantState.Request, response: SetVoiceAssistantState.Response) -> SetVoiceAssistantState.Response:
+        
         request_state: VoiceAssistantState = request.voice_assistant_state
 
-        with self.state_lock:
+        try:
+
+            if self.turning_off: # ignore if currently turning off
+                raise Exception("voice assistant is currently turning off")
```

Comment:
the exception eventually results in an early return, since it is caught in the `except` below. Does the bug that you describe on the newest version of the backend. I also witnessed similar behavior to what you describe, but it should be fixed by now...